File: sil/schedule.py
"""Render and install systemd user units / launchd agents for the worker and web UI.

Every unit's ExecStart points at the shim `~/.local/bin/sil`
(`scripts/sil`, installed here), never at a path inside one plugin version.
A plugin upgrade changes `${CLAUDE_PLUGIN_ROOT}`; the shim resolves the
current install at call time, so an installed schedule never goes stale.
"""

from __future__ import annotations

import shutil
import subprocess
from pathlib import Path

SYSTEMD_WORKER_UNITS = ("sil-worker.service", "sil-worker.timer")
SYSTEMD_WEB_UNIT = "sil-web.service"
LAUNCHD_WORKER_PLIST = "com.kolezka.sil-worker.plist"
LAUNCHD_WEB_PLIST = "com.kolezka.sil-web.plist"


def shim_path() -> Path:
    return Path.home() / ".local" / "bin" / "sil"


def systemd_dir() -> Path:
    return Path.home() / ".config" / "systemd" / "user"


def launchd_dir() -> Path:
    return Path.home() / "Library" / "LaunchAgents"


def render_systemd(interval_min: int, web: bool) -> dict[str, str]:
    shim = str(shim_path())
    units = {
        "sil-worker.service": (
            "[Unit]\n"
            "Description=self-improvement-loop worker, one pass\n\n"
            "[Service]\n"
            "Type=oneshot\n"
            f"ExecStart={shim} worker --once\n"
        ),
        "sil-worker.timer": (
            "[Unit]\n"
            "Description=self-improvement-loop worker schedule\n\n"
            "[Timer]\n"
            "OnBootSec=5min\n"
            f"OnUnitActiveSec={interval_min}min\n\n"
            "[Install]\n"
            "WantedBy=timers.target\n"
        ),
    }
    if web:
        units[SYSTEMD_WEB_UNIT] = (
            "[Unit]\n"
            "Description=self-improvement-loop web UI\n\n"
            "[Service]\n"
            "Type=simple\n"
            f"ExecStart={shim} web\n"
            "Restart=on-failure\n\n"
            "[Install]\n"
            "WantedBy=default.target\n"
        )
    return units


def render_launchd(interval_min: int, web: bool) -> dict[str, str]:
    shim = str(shim_path())
    interval_s = interval_min * 60
    units = {
        LAUNCHD_WORKER_PLIST: (
            '<?xml version="1.0" encoding="UTF-8"?>\n'
            '<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" '
            '"http://www.apple.com/DTDs/PropertyList-1.0.dtd">\n'
            '<plist version="1.0"><dict>\n'
            "  <key>Label</key><string>com.kolezka.sil-worker</string>\n"
            "  <key>ProgramArguments</key><array>\n"
            f"    <string>{shim}</string>\n"
            "    <string>worker</string>\n"
            "    <string>--once</string>\n"
            "  </array>\n"
            f"  <key>StartInterval</key><integer>{interval_s}</integer>\n"
            "  <key>RunAtLoad</key><true/>\n"
            "</dict></plist>\n"
        )
    }
    if web:
        units[LAUNCHD_WEB_PLIST] = (
            '<?xml version="1.0" encoding="UTF-8"?>\n'
            '<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" '
            '"http://www.apple.com/DTDs/PropertyList-1.0.dtd">\n'
            '<plist version="1.0"><dict>\n'
            "  <key>Label</key><string>com.kolezka.sil-web</string>\n"
            "  <key>ProgramArguments</key><array>\n"
            f"    <string>{shim}</string>\n"
            "    <string>web</string>\n"
            "  </array>\n"
            "  <key>KeepAlive</key><true/>\n"
            "  <key>RunAtLoad</key><true/>\n"
            "</dict></plist>\n"
        )
    return units


def _install_shim() -> Path:
    dest = shim_path()
    dest.parent.mkdir(parents=True, exist_ok=True)
    src = Path(__file__).resolve().parent.parent / "scripts" / "sil"
    shutil.copy2(src, dest)
    dest.chmod(0o755)
    return dest


def install(kind: str, interval_min: int = 60, web: bool = False) -> list[Path]:
    """Write the units for `kind` (systemd | launchd) and load them. Returns the
    written file paths. Idempotent: re-running overwrites the same files."""
    _install_shim()
    if kind == "systemd":
        d = systemd_dir()
        d.mkdir(parents=True, exist_ok=True)
        written = []
        for name, content in render_systemd(interval_min, web).items():
            p = d / name
            p.write_text(content, encoding="utf-8")
            written.append(p)
        subprocess.run(["systemctl", "--user", "daemon-reload"], check=False)
        subprocess.run(["systemctl", "--user", "enable", "--now", "sil-worker.timer"], check=False)
        if web:
            subprocess.run(["systemctl", "--user", "enable", "--now", SYSTEMD_WEB_UNIT], check=False)
        return written
    if kind == "launchd":
        d = launchd_dir()
        d.mkdir(parents=True, exist_ok=True)
        written = []
        for name, content in render_launchd(interval_min, web).items():
            p = d / name
            p.write_text(content, encoding="utf-8")
            written.append(p)
            subprocess.run(["launchctl", "load", str(p)], check=False)
        return written
    raise ValueError(f"unknown schedule kind: {kind!r}, use systemd or launchd")


def uninstall(kind: str) -> list[Path]:
    """Stop and remove whatever units of `kind` are installed. A no-op, not an
    error, when nothing is installed."""
    if kind == "systemd":
        d = systemd_dir()
        subprocess.run(["systemctl", "--user", "disable", "--now", "sil-worker.timer"], check=False)
        subprocess.run(["systemctl", "--user", "disable", "--now", SYSTEMD_WEB_UNIT], check=False)
        removed = []
        for name in (*SYSTEMD_WORKER_UNITS, SYSTEMD_WEB_UNIT):
            p = d / name
            if p.exists():
                p.unlink()
                removed.append(p)
        subprocess.run(["systemctl", "--user", "daemon-reload"], check=False)
        return removed
    if kind == "launchd":
        d = launchd_dir()
        removed = []
        for name in (LAUNCHD_WORKER_PLIST, LAUNCHD_WEB_PLIST):
            p = d / name
            if p.exists():
                subprocess.run(["launchctl", "unload", str(p)], check=False)
                p.unlink()
                removed.append(p)
        return removed
    raise ValueError(f"unknown schedule kind: {kind!r}, use systemd or launchd")


def show() -> dict[str, list[str]]:
    """What is currently installed, by kind. Reads the filesystem, runs nothing."""
    d = systemd_dir()
    systemd = sorted(p.name for p in d.glob("sil-*")) if d.exists() else []
    ld = launchd_dir()
    launchd = sorted(p.name for p in ld.glob("com.kolezka.sil-*.plist")) if ld.exists() else []
    return {"systemd": systemd, "launchd": launchd}

File: apps/cli/src/schedule.ts
// Render and install systemd user units / launchd agents for the worker and
// web UI.
//
// Every unit's ExecStart points at the shim `~/.local/bin/sil` (scripts/sil,
// installed here), never at a path inside one plugin version. A plugin
// upgrade changes `${CLAUDE_PLUGIN_ROOT}`; the shim resolves the current
// install at call time, so an installed schedule never goes stale.

import { chmodSync, copyFileSync, existsSync, mkdirSync, readdirSync, rmSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { paths } from "@sil/core";

export const SYSTEMD_WORKER_UNITS = ["sil-worker.service", "sil-worker.timer"] as const;
export const SYSTEMD_WEB_UNIT = "sil-web.service";
export const LAUNCHD_WORKER_PLIST = "com.raqz.sil-worker.plist";
export const LAUNCHD_WEB_PLIST = "com.raqz.sil-web.plist";
/** Labels written by earlier releases; uninstall still removes them. */
export const LEGACY_LAUNCHD_PLISTS = ["com.kolezka.sil-worker.plist", "com.kolezka.sil-web.plist"] as const;
const LAUNCHD_PREFIXES = ["com.raqz.sil-", "com.kolezka.sil-"] as const;

/** Runs a command and never throws; tests inject a fake so no real
 * systemctl/launchctl is ever invoked. */
export type Runner = (cmd: string[]) => void;

export const realRunner: Runner = (cmd) => {
  // Bun.spawnSync throws when the binary is missing, which is the normal case
  // for systemctl on macOS and launchctl on Linux. Uninstall runs both kinds.
  try {
    Bun.spawnSync(cmd, { stdout: "ignore", stderr: "ignore" });
  } catch {
    // Missing tool means nothing of that kind is installed. Nothing to do.
  }
};

// Bun's homedir() does not follow a HOME change made after startup, so a test
// that points HOME at a temp dir would still write into the real LaunchAgents.
function home(): string {
  return process.env["HOME"] || homedir();
}

export function shimPath(): string {
  return join(home(), ".local", "bin", "sil");
}

export function systemdDir(): string {
  return join(home(), ".config", "systemd", "user");
}

export function launchdDir(): string {
  return join(home(), "Library", "LaunchAgents");
}

export function renderSystemd(intervalMin: number, web: boolean): Record<string, string> {
  const shim = shimPath();
  const units: Record<string, string> = {
    "sil-worker.service": [
      "[Unit]",
      "Description=self-improvement-loop worker, one pass",
      "",
      "[Service]",
      "Type=oneshot",
      `ExecStart=${shim} worker --once`,
      "",
    ].join("\n"),
    "sil-worker.timer": [
      "[Unit]",
      "Description=self-improvement-loop worker schedule",
      "",
      "[Timer]",
      "OnBootSec=5min",
      `OnUnitActiveSec=${intervalMin}min`,
      "",
      "[Install]",
      "WantedBy=timers.target",
      "",
    ].join("\n"),
  };
  if (web) {
    units[SYSTEMD_WEB_UNIT] = [
      "[Unit]",
      "Description=self-improvement-loop web UI",
      "",
      "[Service]",
      "Type=simple",
      `ExecStart=${shim} web`,
      // always, not on-failure: `sil web` exits 0 on a plugin update so the
      // restart picks up the new version.
      "Restart=always",
      "RestartSec=2",
      "",
      "[Install]",
      "WantedBy=default.target",
      "",
    ].join("\n");
  }
  return units;
}

// launchd drops stdout and stderr unless told where to put them. Both go to
// the same file `sil logs <name>` reads.
function launchdLogKeys(name: "worker" | "web"): string[] {
  const log = paths.logFile(name);
  return [`  <key>StandardOutPath</key><string>${log}</string>`, `  <key>StandardErrorPath</key><string>${log}</string>`];
}

export function renderLaunchd(intervalMin: number, web: boolean): Record<string, string> {
  const shim = shimPath();
  const intervalS = intervalMin * 60;
  const units: Record<string, string> = {
    [LAUNCHD_WORKER_PLIST]: [
      '<?xml version="1.0" encoding="UTF-8"?>',
      '<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">',
      '<plist version="1.0"><dict>',
      "  <key>Label</key><string>com.raqz.sil-worker</string>",
      "  <key>ProgramArguments</key><array>",
      `    <string>${shim}</string>`,
      "    <string>worker</string>",
      "    <string>--once</string>",
      "  </array>",
      `  <key>StartInterval</key><integer>${intervalS}</integer>`,
      "  <key>RunAtLoad</key><true/>",
      ...launchdLogKeys("worker"),
      "</dict></plist>",
      "",
    ].join("\n"),
  };
  if (web) {
    units[LAUNCHD_WEB_PLIST] = [
      '<?xml version="1.0" encoding="UTF-8"?>',
      '<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">',
      '<plist version="1.0"><dict>',
      "  <key>Label</key><string>com.raqz.sil-web</string>",
      "  <key>ProgramArguments</key><array>",
      `    <string>${shim}</string>`,
      "    <string>web</string>",
      "  </array>",
      "  <key>KeepAlive</key><true/>",
      "  <key>RunAtLoad</key><true/>",
      ...launchdLogKeys("web"),
      "</dict></plist>",
      "",
    ].join("\n");
  }
  return units;
}

function installShim(): string {
  const dest = shimPath();
  mkdirSync(dirname(dest), { recursive: true });
  const src = join(paths.pluginRoot(), "scripts", "sil");
  copyFileSync(src, dest);
  chmodSync(dest, 0o755);
  return dest;
}

/** Pick the supervisor: an explicit flag, else the one this platform has. */
export function resolveKind(opts: { systemd?: boolean; launchd?: boolean }, platform: string = process.platform): "systemd" | "launchd" {
  if (opts.systemd && opts.launchd) throw new Error("sil schedule install takes only one of --systemd or --launchd");
  if (opts.systemd) return "systemd";
  if (opts.launchd) return "launchd";
  if (platform === "darwin") return "launchd";
  if (platform === "linux") return "systemd";
  throw new Error(`no default supervisor on ${platform}, pass --systemd or --launchd`);
}

/** Write the units for `kind` (systemd | launchd) and load them. Returns the
 * written file paths. Idempotent: re-running overwrites the same files. */
export function install(kind: "systemd" | "launchd", intervalMin = 60, web = false, run: Runner = realRunner): string[] {
  installShim();
  if (kind === "systemd") {
    const d = systemdDir();
    mkdirSync(d, { recursive: true });
    const written: string[] = [];
    for (const [name, content] of Object.entries(renderSystemd(intervalMin, web))) {
      const p = join(d, name);
      writeFileSync(p, content, "utf8");
      written.push(p);
    }
    run(["systemctl", "--user", "daemon-reload"]);
    run(["systemctl", "--user", "enable", "--now", "sil-worker.timer"]);
    if (web) run(["systemctl", "--user", "enable", "--now", SYSTEMD_WEB_UNIT]);
    return written;
  }
  if (kind === "launchd") {
    const d = launchdDir();
    mkdirSync(d, { recursive: true });
    // launchd does not create the parent dir of StandardOutPath.
    mkdirSync(dirname(paths.logFile("worker")), { recursive: true });
    const written: string[] = [];
    for (const [name, content] of Object.entries(renderLaunchd(intervalMin, web))) {
      const p = join(d, name);
      writeFileSync(p, content, "utf8");
      written.push(p);
      // load refuses an agent that is already loaded, so a reinstall would
      // keep running the old settings. unload first; it is a no-op otherwise.
      run(["launchctl", "unload", p]);
      run(["launchctl", "load", p]);
    }
    return written;
  }
  throw new Error(`unknown schedule kind: ${JSON.stringify(kind)}, use systemd or launchd`);
}

/** Stop and remove whatever units of `kind` are installed. A no-op, not an
 * error, when nothing is installed. */
export function uninstall(kind: "systemd" | "launchd", run: Runner = realRunner): string[] {
  if (kind === "systemd") {
    const d = systemdDir();
    run(["systemctl", "--user", "disable", "--now", "sil-worker.timer"]);
    run(["systemctl", "--user", "disable", "--now", SYSTEMD_WEB_UNIT]);
    const removed: string[] = [];
    for (const name of [...SYSTEMD_WORKER_UNITS, SYSTEMD_WEB_UNIT]) {
      const p = join(d, name);
      if (existsSync(p)) {
        rmSync(p);
        removed.push(p);
      }
    }
    run(["systemctl", "--user", "daemon-reload"]);
    return removed;
  }
  if (kind === "launchd") {
    const d = launchdDir();
    const removed: string[] = [];
    for (const name of [LAUNCHD_WORKER_PLIST, LAUNCHD_WEB_PLIST, ...LEGACY_LAUNCHD_PLISTS]) {
      const p = join(d, name);
      if (existsSync(p)) {
        run(["launchctl", "unload", p]);
        rmSync(p);
        removed.push(p);
      }
    }
    return removed;
  }
  throw new Error(`unknown schedule kind: ${JSON.stringify(kind)}, use systemd or launchd`);
}

/** What is currently installed, by kind. Reads the filesystem, runs nothing. */
export function show(): { systemd: string[]; launchd: string[] } {
  const d = systemdDir();
  const systemd = existsSync(d) ? readdirSync(d).filter((n) => n.startsWith("sil-")).sort() : [];
  const ld = launchdDir();
  const launchd = existsSync(ld) ? readdirSync(ld).filter((n) => LAUNCHD_PREFIXES.some((pre) => n.startsWith(pre)) && n.endsWith(".plist")).sort() : [];
  return { systemd, launchd };
}
